src/main.rs

```diff
@@ -1,3 +1,204 @@
+        let contents: Rope = (&*params.text_document.text).into();
+        self.per_file.insert(
+            params.text_document.uri.clone(),
+            PerFileState {
+                tree: parse_from_scratch(&contents),
+                contents,
+            },
+        );
+
+        self.run_linting_and_report_diagnostics(&params.text_document.uri)
+            .await;
+    }
+
+    async fn did_change(&self, params: DidChangeTextDocumentParams) {
+        {
+            let mut file_state = self
+                .per_file
+                .get_mut(&params.text_document.uri)
+                .expect("Changed file wasn't loaded");
+            for content_change in &params.content_changes {
+                match content_change.range {
+                    Some(range) => {
+                        let start_char =
+                            lsp_position_to_char_offset(&file_state.contents, range.start);
+                        let end_char = lsp_position_to_char_offset(&file_state.contents, range.end);
+                        let start_byte = file_state.contents.char_to_byte(start_char);
+                        let old_end_byte = file_state.contents.char_to_byte(end_char);
+                        file_state.contents.remove(start_char..end_char);
+                        file_state.contents.insert(start_char, &content_change.text);
+
+                        let new_end_byte = start_byte + content_change.text.len();
+                        let input_edit = InputEdit {
+                            start_byte,
+                            old_end_byte,
+                            new_end_byte,
+                            start_position: position_to_point(range.start),
+                            old_end_position: position_to_point(range.end),
+                            new_end_position: byte_offset_to_point(
+                                &file_state.contents,
+                                new_end_byte,
+                            ),
+                        };
+                        file_state.tree.edit(&input_edit);
+                        file_state.tree = parse(&file_state.contents, Some(&file_state.tree));
+                    }
+                    None => {
+                        file_state.contents = (&*content_change.text).into();
+                        file_state.tree = parse_from_scratch(&file_state.contents);
+                    }
+                }
+            }
+        }
+
+        self.run_linting_and_report_diagnostics(&params.text_document.uri)
+            .await;
+    }
+}
+
+#[derive(Debug)]
+struct PerFileState {
+    contents: Rope,
+    tree: Tree,
+}
+
+fn parse_from_scratch(contents: &Rope) -> Tree {
+    parse(contents, None)
+}
+
+fn parse(contents: &Rope, old_tree: Option<&Tree>) -> Tree {
+    let mut parser = Parser::new();
+    parser
+        .set_language(SupportedLanguage::Rust.language())
+        .unwrap();
+    contents.parse(&mut parser, old_tree).unwrap()
```

Comment:
This is leveraging the `impl Parseable for Rope` defined in `tree_sitter_grep` (?)